File: .opencode/tools/memory.ts
import { tool } from "@opencode-ai/plugin"
import {
  assertSafeEntry,
  findSubstringMatches,
  formatStore,
  MEMORY_TARGETS,
  parseTarget,
  readEntries,
  usage,
  wouldExceedLimit,
  writeEntries,
} from "../lib/memory-store"

type Action = "add" | "replace" | "remove"

const ACTIONS = new Set(["add", "replace", "remove"])

function parseAction(value: string): Action {
  const action = value.trim().toLowerCase()

  if (!ACTIONS.has(action)) {
    throw new Error(`action must be "add", "replace", or "remove", got: ${value}`)
  }

  return action as Action
}

function requireContent(value: string | undefined) {
  const content = value?.trim() ?? ""

  if (!content) {
    throw new Error("content is required for add and replace actions")
  }

  assertSafeEntry(content)
  return content
}

function currentEntriesText(entries: string[]) {
  return entries.length > 0
    ? entries.map((entry, index) => `${index + 1}. ${entry}`).join("\n")
    : "(empty)"
}

function capacityError(action: Action, target: keyof typeof MEMORY_TARGETS, entries: string[], nextEntries: string[]) {
  const currentUsage = usage(target, entries)
  const nextUsage = usage(target, nextEntries)

  return {
    output: `${MEMORY_TARGETS[target].label} is at ${currentUsage.used}/${currentUsage.limit} chars. ${action} would use ${nextUsage.used}/${nextUsage.limit} chars and exceed the limit. Replace, remove, or consolidate existing entries first.\nCurrent entries:\n${currentEntriesText(entries)}`,
    metadata: {
      success: false,
      error: "capacity_exceeded",
      target,
      usage: currentUsage,
      nextUsage,
      current_entries: entries,
    },
  }
}

function substringError(target: keyof typeof MEMORY_TARGETS, oldText: string, matches: number, entries: string[]) {
  const problem = matches === 0
    ? `No ${target} memory entry matched old_text: ${oldText}`
    : `${matches} ${target} memory entries matched old_text: ${oldText}. Use a more specific substring.`

  return {
    output: `${problem}\nCurrent entries:\n${currentEntriesText(entries)}`,
    metadata: {
      success: false,
      error: matches === 0 ? "not_found" : "ambiguous_match",
      target,
      old_text: oldText,
      current_entries: entries,
    },
  }
}

function success(action: Action, target: keyof typeof MEMORY_TARGETS, entries: string[], message: string) {
  return {
    output: `${message}\n${formatStore(target, entries)}`,
    metadata: {
      success: true,
      action,
      target,
      usage: usage(target, entries),
    },
  }
}

export default tool({
  description: "Manage bounded persistent memory stores with Hermes-style add, replace, and remove actions.",
  args: {
    action: tool.schema.string().describe("One of: add, replace, remove"),
    target: tool.schema.string().optional().describe("memory for agent notes, or user for user profile. Defaults to memory."),
    content: tool.schema.string().optional().describe("Memory entry content for add or replace actions"),
    old_text: tool.schema.string().optional().describe("Unique substring identifying the entry to replace or remove"),
  },
  async execute(args, context) {
    const action = parseAction(args.action)
    const target = parseTarget(args.target)
    const entries = await readEntries(context, target)

    if (action === "add") {
      const content = requireContent(args.content)

      if (entries.includes(content)) {
        return success(action, target, entries, `Duplicate ${target} memory not added.`)
      }

      const nextEntries = [...entries, content]
      if (wouldExceedLimit(target, nextEntries)) {
        return capacityError(action, target, entries, nextEntries)
      }

      await writeEntries(context, target, nextEntries)
      return success(action, target, nextEntries, `Added ${target} memory.`)
    }

    const oldText = args.old_text?.trim() ?? ""
    const matches = findSubstringMatches(entries, oldText)
    if (matches.length !== 1) {
      return substringError(target, oldText, matches.length, entries)
    }

    const [{ index, entry }] = matches

    if (action === "remove") {
      const nextEntries = entries.filter((_, entryIndex) => entryIndex !== index)

      await writeEntries(context, target, nextEntries)
      return success(action, target, nextEntries, `Removed ${target} memory: ${entry}`)
    }

    const content = requireContent(args.content)
    const duplicateIndex = entries.findIndex((current, entryIndex) => entryIndex !== index && current === content)
    if (duplicateIndex !== -1) {
      return success(action, target, entries, `Replacement already exists in ${target} memory; no duplicate added.`)
    }

    const nextEntries = entries.map((current, entryIndex) => (entryIndex === index ? content : current))
    if (wouldExceedLimit(target, nextEntries)) {
      return capacityError(action, target, entries, nextEntries)
    }

    await writeEntries(context, target, nextEntries)
    return success(action, target, nextEntries, `Replaced ${target} memory.`)
  },
})

File: tools/memory.py
from __future__ import annotations

import re

from .common import ENTRY_SEPARATOR, WORKSPACE, json_result


MEMORY_DIR = WORKSPACE / ".opencode" / "memories"
MEMORY_TARGETS = {
    "memory": {"file": "MEMORY.md", "label": "MEMORY (agent notes)", "limit": 2200},
    "user": {"file": "USER.md", "label": "USER PROFILE", "limit": 1375},
}


def _memory_path(target: str):
    return MEMORY_DIR / MEMORY_TARGETS[target]["file"]


def _parse_target(value: str | None) -> str:
    target = (value or "memory").strip().lower()
    if target not in MEMORY_TARGETS:
        raise ValueError(f'target must be either "memory" or "user", got: {value}')
    return target


def _parse_entries(text: str) -> list[str]:
    return [entry.strip().replace("\r\n", "\n") for entry in re.split(rf"\s*{ENTRY_SEPARATOR}\s*", text) if entry.strip()]


def _read_entries(target: str) -> list[str]:
    path = _memory_path(target)
    if not path.exists():
        return []
    return _parse_entries(path.read_text(encoding="utf-8"))


def _write_entries(target: str, entries: list[str]) -> None:
    MEMORY_DIR.mkdir(parents=True, exist_ok=True)
    normalized = [entry.strip().replace("\r\n", "\n") for entry in entries if entry.strip()]
    content = f"\n{ENTRY_SEPARATOR}\n".join(normalized)
    _memory_path(target).write_text(f"{content}\n" if content else "", encoding="utf-8")


def _usage(target: str, entries: list[str]) -> dict[str, int]:
    used = len(ENTRY_SEPARATOR.join(entries))
    limit = int(MEMORY_TARGETS[target]["limit"])
    return {"used": used, "limit": limit, "percent": round((used / limit) * 100)}


def _format_store(target: str, entries: list[str]) -> str:
    config = MEMORY_TARGETS[target]
    usage = _usage(target, entries)
    body = ENTRY_SEPARATOR.join(entries) if entries else "(empty)"
    return f"{config['label']} [{usage['percent']}% - {usage['used']}/{usage['limit']} chars]\n{body}"


def get_memories() -> str:
    """Read the session-start snapshot of bounded persistent memory stores."""
    stores = [_format_store(target, _read_entries(target)) for target in ("memory", "user")]
    return json_result("\n\n".join(stores))


def _assert_safe_memory(content: str) -> None:
    if ENTRY_SEPARATOR in content:
        raise ValueError(f"memory content cannot contain the {ENTRY_SEPARATOR} delimiter")
    blocked_patterns = [
        r"ignore\s+(all\s+)?(previous|prior|above)\s+(instructions|system|developer)",
        r"(reveal|print|dump|exfiltrate).{0,40}(system prompt|developer message|secrets?|credentials?)",
        r"(?:password|token|secret|api[_ -]?key)\s*[:=]\s*[\"']?[A-Za-z0-9_./+=-]{12,}",
        r"-----BEGIN [A-Z ]*PRIVATE KEY-----",
    ]
    if any(re.search(pattern, content, re.IGNORECASE) for pattern in blocked_patterns):
        raise ValueError("memory content matched a blocked injection or secret pattern")


def memory(action: str, target: str | None = None, content: str | None = None, old_text: str | None = None) -> str:
    """Manage bounded persistent memory stores with add, replace, and remove actions."""
    action = action.strip().lower()
    if action not in {"add", "replace", "remove"}:
        raise ValueError(f'action must be "add", "replace", or "remove", got: {action}')

    parsed_target = _parse_target(target)
    entries = _read_entries(parsed_target)

    def success(next_entries: list[str], message: str) -> str:
        return json_result(
            f"{message}\n{_format_store(parsed_target, next_entries)}",
            {"success": True, "action": action, "target": parsed_target, "usage": _usage(parsed_target, next_entries)},
        )

    if action == "add":
        new_content = (content or "").strip()
        if not new_content:
            raise ValueError("content is required for add and replace actions")
        _assert_safe_memory(new_content)
        if new_content in entries:
            return success(entries, f"Duplicate {parsed_target} memory not added.")
        next_entries = [*entries, new_content]
        if _usage(parsed_target, next_entries)["used"] > MEMORY_TARGETS[parsed_target]["limit"]:
            return json_result("Memory capacity would be exceeded.", {"success": False, "error": "capacity_exceeded"})
        _write_entries(parsed_target, next_entries)
        return success(next_entries, f"Added {parsed_target} memory.")

    needle = (old_text or "").strip()
    if not needle:
        raise ValueError("old_text must not be empty")
    matches = [(index, entry) for index, entry in enumerate(entries) if needle in entry]
    if len(matches) != 1:
        return json_result(
            f"Expected exactly one {parsed_target} memory entry matching old_text; found {len(matches)}.",
            {"success": False, "error": "not_found" if not matches else "ambiguous_match", "current_entries": entries},
        )

    index, old_entry = matches[0]
    if action == "remove":
        next_entries = [entry for entry_index, entry in enumerate(entries) if entry_index != index]
        _write_entries(parsed_target, next_entries)
        return success(next_entries, f"Removed {parsed_target} memory: {old_entry}")

    new_content = (content or "").strip()
    if not new_content:
        raise ValueError("content is required for add and replace actions")
    _assert_safe_memory(new_content)
    next_entries = [new_content if entry_index == index else entry for entry_index, entry in enumerate(entries)]
    if _usage(parsed_target, next_entries)["used"] > MEMORY_TARGETS[parsed_target]["limit"]:
        return json_result("Memory capacity would be exceeded.", {"success": False, "error": "capacity_exceeded"})
    _write_entries(parsed_target, next_entries)
    return success(next_entries, f"Replaced {parsed_target} memory.")
